File: Linked List/86. Partition List/86. Partition List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        # 1. 处理特殊情况，即head==None的情况
        if not head:
            return head

        # 2. 创建一个比给定值的小的结点的链表low和比给定值大的结点的链表high，同时创建游标指针lowpos和highpos
        pos=head          # 原始链表的游标指针
        low=ListNode(0)   # 给定值的小的结点的链表low
        high=ListNode(0)  # 比给定值大的结点的链表high

        lowpos=low        # 链表low的游标指针
        highpos=high      # 链表high的游标指针

        while pos:
            if pos.val<x:
                lowpos.next=ListNode(pos.val)
                lowpos=lowpos.next
            
            else:
                highpos.next=ListNode(pos.val)
                highpos=highpos.next
            
            pos=pos.next
        
        lowpos.next=high.next
        return low.next

File: Linked List/86. Partition List/86. Partition List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

//较为简单的一题，创建两个临时链表low和high，然后使用一个游标指针遍历原始链表，若当前链表结点小于值x，则将该链表结点复制一份加入链表low，否则将该链表结点复制一份加入链表high

class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        // 1. 处理原始链表为空的情况
        if(!head)        
        return head;

        // 2. 创建一个游标指针pos以及两个临时链表low和high
        ListNode *pos=head;
        ListNode *low=new ListNode(0);     //含有哑结点
        ListNode *high=new ListNode(0);    //含有哑结点
        ListNode *lowpos=low;
        ListNode *highpos=high;

        // 3. 使用游标指针pos遍历整个原始链表，将小于x的结点加入low，将大于等于x的结点加入high
        while(pos)
        {
            if(pos->val<x)
            {
                lowpos->next=new ListNode(pos->val);
                lowpos=lowpos->next;
            }
            else
            {
                highpos->next=new ListNode(pos->val);
                highpos=highpos->next;
            }

            pos=pos->next;
        }

        // 4. 然后将high链表连接在low后，需要注意特殊情况，low链表可能为空
        if(!high)
        return low;
        else if(!low)
        return high;
        else
        {
            lowpos->next=high->next;
            delete high;          //删除哑结点
            ListNode *temp=low;   
            low=low->next;
            delete temp;          //删除哑结点

            return low;
        }
    }
};